spec/validators/waste_exemptions_engine/postcode_validator_spec.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+module Test
+  PostcodeValidatable = Struct.new(:postcode) do
+    include ActiveModel::Validations
+
+    validates :postcode, "waste_exemptions_engine/postcode": true
+  end
+end
+
+module WasteExemptionsEngine
+  RSpec.describe PostcodeValidator, type: :model do
+    let(:valid_postcode) { "BS1 5AH" }
+    let(:invalid_postcode) { "foo" }
+    let(:postcode_without_addresses) { "AA1 1AA" }
+
+    subject(:validatable) { Test::PostcodeValidatable.new(valid_postcode) }
+
+    describe "#validate_each" do
+      context "when the postcode is valid" do
+        before(:context) { VCR.insert_cassette("postcode_valid") }
+        after(:context) { VCR.eject_cassette }
+
+        it "gets called" do
+          expect_any_instance_of(PostcodeValidator)
+            .to receive(:validate_each)
+            .once
+
+          validatable.valid?
+        end
+
+        it "the errors are empty" do
+          validatable.valid?
+          expect(validatable.errors).to be_empty
+        end
+      end
+
+      context "when the postcode is not valid" do
+        subject(:validatable) { Test::PostcodeValidatable.new(invalid_postcode) }
+
+        it "adds a validation error to the record" do
+          validatable.valid?
+          expect(validatable.errors[:postcode].count).to eq(1)
+        end
+
+        context "because the postcode is not present" do
+          subject(:validatable) { Test::PostcodeValidatable.new }
+
+          it "adds an appropriate validation error" do
+            validatable.valid?
+            expect(validatable.errors[:postcode]).to eq(["Enter a postcode"])
```

Comment:
I'd advocate referring back to the locale rather than hard coding the expected text in the test.

Experience on digital services is that content changes frequently so hard coding tends to make things quite brittle. However this one probably needs more of a chat. This might be my lack of experience showing through 😁

Reply:
Fixed in the latest commit.